### ISSUE TITLE
Issue warning on overloading variables?

### ISSUE THREAD

**Issue by kintel** [openscad/openscad#340]
Alan Cox suggests on the mailing list that we should issue a warning on attempts to overload a variable, as it would shadow an existing definition and potentially cause unintended behavior.

Simple example which would in this case cause a warning. In more complex scenarios it would be harder to spot the problem (e.g. overriding variable in an outer/global scope):

```
module mymodule(myvar) {
  myvar = 3;
  sphere(myvar);
}

mymodule(6);
```
##

**Comment by kintel** [comment]
@MichaelPFrey Does your warning work fully or partially fix this?

**Comment by MichaelPFrey** [comment]
@kintel No, not really and I am kind of supprised how granular the scopes are.
My warning only triggers, when a variable is overwritten in the same scope.
The parameters of a function and the function body are different scopes.

To create a warning for
```
module mymodule(myvar) {
  myvar = 3;
  sphere(myvar);
}

mymodule(6);
```
I could implement a check one scope "below" the current one, but then I need to know what kind of scope it is (which has to be implemented) before doing that. Sounds tricky and iterating over two scopes does not sound smart.

Alternatively, we could to use the same scope for the function parameter and function body.
I do not know how hard that would be and if that would have unwanted side effects.

I will look into it.

**Comment by MichaelAtOz** [comment]
> Alternatively, we could to use the same scope for the function parameter and function body.
I do not know how hard that would be and if that would have unwanted side effects.

Tread carefully.

```
module mymodule(myvar,myvar2=myvar) {
  myvar = 3;
  sphere(myvar);
  echo("mymodule: ",myvar=myvar,myvar2=myvar2); // 3, 9
}

myvar=9;
mymodule(6);
echo("main: ",myvar=myvar); // 9
```

**Comment by MichaelPFrey** [comment]
To address this:
```
function test_function() = a;
b=test_function(a=3);
a=4;
echo(b);
```
we need separate scopes for the function parameter and function body - so that makes the decision easier.

**Comment by MichaelPFrey** [comment]
closed with #2628?

In theory, it could be improved, but it is tricky to warn every possible issue, when mean while
```
function clamp(a) = a > 10 ? 10 : a < 0 ? 0: a;

module mytest(a=5){
    a=clamp(a);
    echo(a);
}

mytest(20);
```
is acceptable. Warning, when a parameter is overwritten by a literal is a decent compromise.
